solr/core/src/java/org/apache/solr/cloud/overseer/CollectionMutator.java

```diff
@@ -113,10 +114,13 @@ public ZkWriteCommand modifyCollection(final ClusterState clusterState, ZkNodePr
           log.error("trying to set perReplicaState to {} from {}", val, coll.isPerReplicaState());
           continue;
         }
-        replicaOps = PerReplicaStatesOps.modifyCollection(coll, enable, PerReplicaStates.fetch(coll.getZNode(), zkClient, null));
+        PerReplicaStates prs = PerReplicaStates.fetch(coll.getZNode(), zkClient, null);
+        replicaOps = PerReplicaStatesOps.modifyCollection(coll, enable, prs);
+        if(!enable) {
+          coll = updateReplicasFromPrs(coll, prs);
+        }
```

Comment:
Is there still a race condition here? What happens if a replica goes down (or changes leadership) between the fetch and the persist which happens much later?